### PR TITLE
Startup IP Followup

### LINKED ISSUE #2529: Ursula dies at startup while learning external ip address when mainnet.nucypher.network is down
**Describe the Bug**
Ursula worker dies at startup due to unhandled exception.

From the traceback below, it appears that while handling a failed connection, an attempt to determine the external ip address by connecting to the default teacher is made which also throws.

At the time this happened, the site https://mainnet.nucypher.network:9151 appeared to be down.

**To Reproduce**
Suggest ensuring https://mainnet.nucypher.network:9151 is down and then attempt starting Ursula.

If bringing https://mainnet.nucypher.network:9151 down intentionally is undesirable, suggest adding local iptables / nftables rule to reject connections.

**Traceback or Screenshots (Optional)**
```
Verbose mode is enabled
Authenticating Ursula
Qualifying worker
✓ Worker is bonded to 0x559A0408aAf3beAb55E80Bd5E2ea83D233f43F7d
✓ Worker is funded with 0.378943746597798709 ETH
Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/nucypher/characters/lawful.py", line 1466, in __attempt
    teacher = cls.from_seed_and_stake_info(seed_uri=teacher_uri,
  File "/usr/lib/python3.8/site-packages/nucypher/characters/lawful.py", line 1507, in from_seed_and_stake_info
    certificate = network_middleware.get_certificate(host=host, port=port)
  File "/usr/lib/python3.8/site-packages/nucypher/network/middleware.py", line 176, in get_certificate
    seednode_certificate = ssl.get_server_certificate(addr=(host, port))
  File "/usr/lib/python3.8/ssl.py", line 1483, in get_server_certificate
    with  create_connection(addr) as sock:
  File "/usr/lib/python3.8/socket.py", line 808, in create_connection
    raise err
  File "/usr/lib/python3.8/socket.py", line 796, in create_connection
    sock.connect(sa)
ConnectionRefusedError: [Errno 111] Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/bin/nucypher", line 8, in <module>
    sys.exit(nucypher_cli())
  File "/usr/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/usr/lib/python3.8/site-packages/nucypher/cli/options.py", line 169, in wrapper
    return func(**kwargs)
  File "/usr/lib/python3.8/site-packages/nucypher/cli/options.py", line 169, in wrapper
    return func(**kwargs)
  File "/usr/lib/python3.8/site-packages/nucypher/cli/options.py", line 169, in wrapper
    return func(**kwargs)
  File "/usr/lib/python3.8/site-packages/nucypher/cli/commands/ursula.py", line 390, in run
    perform_startup_ip_check(emitter=emitter, ursula=URSULA, force=force)
  File "/usr/lib/python3.8/site-packages/nucypher/cli/actions/configure.py", line 157, in perform_startup_ip_check
    external_ip = determine_external_ip_address(network=ursula.domain, known_nodes=ursula.known_nodes)
  File "/usr/lib/python3.8/site-packages/nucypher/utilities/networking.py", line 172, in determine_external_ip_address
    rest_host = get_external_ip_from_default_teacher(network=network)
  File "/usr/lib/python3.8/site-packages/nucypher/utilities/networking.py", line 105, in get_external_ip_from_default_teacher
    teacher = Ursula.from_teacher_uri(teacher_uri=top_teacher_url,
  File "/usr/lib/python3.8/site-packages/nucypher/characters/lawful.py", line 1481, in from_teacher_uri
    return __attempt()
  File "/usr/lib/python3.8/site-packages/nucypher/characters/lawful.py", line 1477, in __attempt
    return __attempt(attempt=attempt + 1)
  File "/usr/lib/python3.8/site-packages/nucypher/characters/lawful.py", line 1463, in __attempt
    raise ConnectionRefusedError("Host {} Refused Connection".format(teacher_uri))
ConnectionRefusedError: Host https://mainnet.nucypher.network:9151 Refused Connection
```

**System (please complete the following information):**
 - OS Version: Alpine Linux 3.12 Container / Kernel: Linux 4.18.0-193.19.1.el8_2.x86_64 unknown
 - Nucypher Version: 4.5.2
 - Ethereum Node Version: geth 1.9.25-stable-e7872729

### PR DESCRIPTION
**Pull request by KPrasch**
**Type of PR:**

- [X] Bugfix
- [ ] Feature
- [ ] Documentation
- [X] Other (supporting tests)

**Required reviews:** 

- [ ] 1
- [X] 2
- [ ] 3

**Issues Fixed**

Closes #2529 
Follow-up for PR #2462 

**What this does:**

- Bugfix: Handle default teacher unavailability during IP startup check
- Bugfix: Fixes a bug when an empty `FleetSensor` instance is passed into `random.sample`
- Bugfix: Enforces a minimum number of peers for fleet-sourced IP determination
- Bugfix: Use `NucypherMiddlewareClient` for peer requests to `/ping`
- Deeper unit and integrations tests for automated IP determination

**Notes for Reviewers**

- I suggest a bugfix release after merging (patch)
- Staged on Ibex nodes:

```
2021-01-17T15:29:25-0800 [external-ip-detection#info] Fetched external IP address (1.2.3.4) from default teacher (2.4.6.8:9151).
```

### CODE REVIEW

File: nucypher/utilities/networking.py [hunk 1]
```diff
@@ -145,10 +158,10 @@ def get_external_ip_from_known_nodes(known_nodes: FleetSensor,
     if len(known_nodes) < sample_size:
         return  # There are too few known nodes
     sample = random.sample(list(known_nodes), sample_size)
+    client = NucypherMiddlewareClient()
     for node in sample:
-        ip = _request(url=node.rest_url())
+        ip = _request_from_node(teacher=node, client=client)
         if ip:
-            log.info(f'Fetched external IP address ({ip}) from randomly selected known node(s).')
```
Comment (cygnusv) [inline nucypher/utilities/networking.py:142]: I think log messages were potentially useful. Any particular reason for removing them?

**Review by cygnusv — APPROVED** [review]

File: nucypher/utilities/networking.py (hunk 1 above)
Comment (KPrasch) [reply nucypher/utilities/networking.py:142]: Hmm, unintentionally removed in this case.  I agree they are useful and want to restore it.

**Review by vepkenez — APPROVED** [review]

**Review by derekpierre — CHANGES_REQUESTED** [review]
Feels like we can extract a newsfragment from these 😄  :

> Bugfix: Handle default teacher unavailability during IP startup check
Bugfix: Fixes a bug when an empty FleetSensor instance is passed into random.sample
Bugfix: Enforces a minimum number of peers for fleet-sourced IP determination
Bugfix: Use NucypherMiddlewareClient for peer requests to /ping

`Ensure minimum number of available peers for fleet-sourced IP determination and better handle default teacher unavailability scenarios on startup`

File: nucypher/utilities/networking.py (hunk 1 above)
Comment (KPrasch) [reply nucypher/utilities/networking.py:142]: :heavy_check_mark:

File: nucypher/utilities/networking.py [hunk 2]
```diff
@@ -76,11 +80,41 @@ def __request(url: str, certificate=None) -> Union[str, None]:
...
+        try:
+            ip = str(ip_address(response.text))
+        except ValueError:
+            error = f'Teacher at {teacher.rest_url()} returned an invalid IP response; Got {response.text}'
+            raise UnknownIPAddress(error)
+        log.info(f'Fetched external IP address ({ip}) from teacher ({teacher.rest_url()}).')
+        return ip
+    else:
+        # Something strange happened... move on anyways.
+        log.debug(f'Failed to get external IP from teacher node ({response.status_code})')
```
Comment (derekpierre) [inline nucypher/utilities/networking.py:108]:
```suggestion
        log.debug(f'Failed to get external IP from teacher node ({teacher.rest_url()} returned {response.status_code})')
```

**Review by derekpierre — APPROVED** [review]
🎸 - looks good! One minor suggestion.

File: nucypher/utilities/networking.py (hunk 2 above)
Comment (KPrasch) [reply nucypher/utilities/networking.py:108]: I'm wondering if this will still work for a `NodeSprout`?

Comment (derekpierre) [reply nucypher/utilities/networking.py:108]: Interesting. Do we have that same problem for the other uses of `teacher.rest_url()` in lines above? i.e. when `response.status_code` is 200.

Comment (KPrasch) [reply nucypher/utilities/networking.py:108]: :heavy_check_mark:

Comment (derekpierre) [reply nucypher/utilities/networking.py:108]: Updated suggestion

```suggestion
        log.debug(f'Failed to get external IP from teacher node ({teacher} returned {response.status_code})')
```

**Pull request merged**